hazelcast/src/main/java/com/hazelcast/internal/management/ManagementCenterService.java

```diff
@@ -46,36 +48,130 @@
+        private final LongSupplier clock;
+        private volatile long lastMCEventsPollMillis;
+        private volatile long lastCleanupTimestamp;
+        private final ConcurrentMap<Address, LastPollRecord> lastPollRecordPerMCAddress = new ConcurrentHashMap<>();
+        private final ConcurrentArrayRingbuffer<Event> mcEvents;
+
+        MCEventStore(LongSupplier clock, ConcurrentArrayRingbuffer<Event> mcEvents) {
+            this.clock = clock;
+            this.lastMCEventsPollMillis = clock.getAsLong();
+            this.lastCleanupTimestamp = this.lastMCEventsPollMillis;
+            this.mcEvents = mcEvents;
+        }
+
+        void log(Event event) {
+            if (clock.getAsLong() - lastMCEventsPollMillis > MC_EVENTS_WINDOW_MILLIS) {
+                // ignore event and clear the queue if the last poll happened a while ago
+                onMCEventWindowExceeded();
+            } else {
+                mcEvents.add(event);
+                cleanUpLastAccessRecords();
+            }
+        }
+
+        private static boolean isOutOfTimeWindow(long nowInMillis, long subjectTimestampInMillis) {
+            return nowInMillis - subjectTimestampInMillis > MC_EVENTS_WINDOW_MILLIS;
+        }
+
+        /**
+         * Removing {@link #lastPollRecordPerMCAddress} entries older than {@link #MC_EVENTS_WINDOW_MILLIS} according to their
+         * {@code lastAccessTime}.
+         * <p>
+         * No-op if less than {@link #MC_EVENTS_WINDOW_MILLIS} millis passed since the last cleanup. This avoids too frequent
+         * unnecessary iterations on the {@link #lastPollRecordPerMCAddress} entries.
+         */
+        private void cleanUpLastAccessRecords() {
+            long now = clock.getAsLong();
+            if (!isOutOfTimeWindow(now, lastCleanupTimestamp)) {
+                return;
+            }
+            lastPollRecordPerMCAddress.entrySet().removeIf(entry -> isOutOfTimeWindow(now, entry.getValue().lastAccessTime));
+            lastCleanupTimestamp = now;
+        }
+
+        void onMCEventWindowExceeded() {
+            mcEvents.clear();
+            lastPollRecordPerMCAddress.clear();
+        }
+
+        /**
+         * @param mcRemoteAddr the address of the calling MC instance.
+         * @return the events which were added to the queue since this MC last polled, or all known events if the MC polls for the
+         * first time.
+         */
+        public List<Event> pollMCEvents(Address mcRemoteAddr) {
+            lastMCEventsPollMillis = clock.getAsLong();
+            LastPollRecord lastPollRecord = lastPollRecordPerMCAddress.get(mcRemoteAddr);
+            long sequence;
+            if (lastPollRecord == null) {
+                sequence = 0;
+            } else {
+                sequence = lastPollRecord.nextSequence;
+            }
+            ConcurrentArrayRingbuffer.RingbufferSlice<Event> slice = mcEvents.copyFrom(sequence);
```

Comment:
`mcEvents.copyFrom(sequence)` may throw `IllegalArgumentException` (see `ConcurrentArrayRingbuffer#checkSequence()`). We should handle this exception here.